### PR TITLE
fix(core): harden queued load authorization

### PR DESCRIPTION
**Pull request by robotoer**
## Summary

- retain the exact signed load-response bytes and reauthorize queued writer/bootstrap responses after preceding membership mutations
- stage bootstrap document and keychain state, then expose it only after all authorization, fetch, decode, and persistence work succeeds
- enforce exact data-property snapshots for untrusted bootstrap bundles and reject accessors, extra properties, lookalike buffers, and shared buffers
- propagate cancellation through invitation catch-up and use a bounded, signal-aware missing-block fetch pool with aggregate/per-block limits
- make final completion and observer notification non-yielding so timed-out loads cannot publish a late success
- redact arbitrary provider errors and cover partial, no-op, retry, timeout, malformed-input, and concurrency paths

## Security rationale

A load response may wait behind membership work after its first authorization decision. Rechecking in queue order binds the accepted response to the current ACL, while transactional staging prevents partially applied or late bootstrap state from becoming observable. Exact snapshots and bounded fetches also remove accessor/TOCTOU and resource-exhaustion paths at the untrusted load boundary.

The stricter byte budgets apply to initial/bootstrap loads and invitation catch-up; ordinary synchronization retains its existing compatibility behavior.

## Validation

- core: 90 suites, 1,892 tests
- automerge: 3 suites, 147 tests
- yjs: 4 suites, 164 tests
- core, Automerge, and Yjs TypeScript builds
- seven-commit range-diff preserved exactly across the ACL-stack rebase
- adversarial review and `git diff --check`

Part of #186.

### CODE REVIEW

**Comment by robotoer** [comment]
@copilot review

File: packages/core/src/peerborne-document.ts [hunk 1]
```diff
@@ -1525,18 +1952,22 @@ export class PeerborneDocument<
...
   }
 
+  private _schedulePendingWelcomeDrain(): void {
+    if (this._pendingWelcomes.size === 0) return;
+    void this._runStateMutation(() => this._drainPendingWelcomesUnlocked())
+      .catch((err) => {
+        console.error(
+          `Failed to drain pending BeeKEM Welcomes for ${this.documentPath}:`,
+          err,
+        );
```
[the remark this answers is not in the document]
Comment (robotoer) [reply packages/core/src/peerborne-document.ts]: The current _schedulePendingWelcomeDrain catch has no error parameter and logs only a fixed context string. No caught provider/decryption value is passed to the logger. This earlier version of the finding is addressed.

File: packages/core/src/peerborne-document.ts [hunk 2]
```diff
@@ -2336,6 +2768,7 @@ export class PeerborneDocument<
       console.log('Sending encrypted shared doc-load response');
 
       if (!isSharedProtocolHandlerActive(admission)) return;
+      this._assertBootstrapResponseRevision(bootstrapRevision);
```
[the remark this answers is not in the document]
Comment (robotoer) [reply packages/core/src/peerborne-document.ts]: All three response handlers now use _sendAuthorizedLoadResponse. That helper runs inside _runStateMutation, calls _isLoadRequesterAuthorized(message, false) against the current ACL, and invokes stream.sink while retaining the membership queue. The bootstrap revision check is no longer the sole final authorization gate.

File: packages/core/src/peerborne-document.ts [hunk 3]
```diff
@@ -5697,102 +6574,146 @@ export class PeerborneDocument<
...
+        await this._assertAcceptedInvitationMembership(issuerPublicKey, role);
 
-    this._invitationEpoch = new Uint8Array(bundle.welcomeEpochId);
-    this._invitationBootstrapReady = true;
+        this._invitationEpoch = new Uint8Array(
+          invitationBundle.welcomeEpochId,
+        );
+        this._invitationBootstrapReady = true;
+        try {
+          await this._completeBootstrapStateApplicationUnlocked();
```
[the remark this answers is not in the document]
Comment (robotoer) [reply packages/core/src/peerborne-document.ts]: _activateAcceptedInvitationBootstrap now keeps the application pending through _open, issuer-pinned catch-up, and the final membership assertion. It completes only after those gates succeed. Its catch restores the pending marker before closing transport state, so the failed instance remains discard-only.

File: packages/core/src/peerborne-document.ts [hunk 4]
```diff
@@ -1277,73 +1646,125 @@ export class PeerborneDocument<
...
+            if (signal?.aborted) throwIfLoadAborted(signal);
+            if (fetchLimitExceeded && fetchController?.signal.aborted) return;
+            if (enforceFetchLimits && error instanceof RangeError) {
+              fetchLimitExceeded = true;
+              if (!fetchController?.signal.aborted) {
+                fetchController?.abort(
+                  new RangeError('Missing change block fetch limits exceeded'),
+                );
+              }
+              return;
```
[the remark this answers is not in the document]
Comment (robotoer) [reply packages/core/src/peerborne-document.ts]: The current implementation uses a dedicated private _LoadFetchLimitExceededError in both missing-block workers and quorum prefetch. Arbitrary provider RangeError values take the ordinary redacted missing-block failure path. The load-boundary suite includes coverage for that distinction.

File: packages/core/src/peerborne-document.ts [hunk 5]
```diff
@@ -5790,121 +6831,151 @@ export class PeerborneDocument<
...
-    const epochPresent = hydratedKeys.some(([keyId]) =>
-      constantTimeEqual(keyId, bundle.welcomeEpochId),
-    );
-    if (!epochPresent || !this._keychain.getKey(bundle.welcomeEpochId)) {
-      throw new Error('Invitation Welcome did not install its advertised epoch key');
-    }
+    await this._runInvitationBootstrapStateApplication(
+      async () => {
+        this._keychain.merge(keychainChanges);
+        const hydratedKeys = await this._keychain.keys();
```
[the remark this answers is not in the document]
Comment (robotoer) [reply packages/core/src/peerborne-document.ts]: acceptInvitationBootstrap now requires prepareMerge, hydrates and decrypts against the detached projection, verifies the inner issuer signature and sealed/signed keychain binding, completes the required CID sync, and checks the initial membership topology before preparedKeychainMerge.commit(). The invitation-transaction tests cover signature, completeness, membership, and mismatched-delta rejection without live keychain mutation.

File: packages/core/src/peerborne-document.ts [hunk 6]
```diff
@@ -2786,9 +3344,25 @@ export class PeerborneDocument<
...
+    const aggregateBudget: DocumentChangeFetchBudget = {
+      maxBytes: responseLimit,
+      consumedBytes: 0,
+    };
+    const prefetchedBlocks = new Map<string, Uint8Array>();
+    const changeFetchOptions: DocumentChangeFetchOptions = {
+      signal,
+      maxBlockBytes: responseLimit,
+      maxAggregateBlockBytes: responseLimit,
+      aggregateBudget,
```
[the remark this answers is not in the document]
Comment (robotoer) [reply packages/core/src/peerborne-document.ts]: Addressed in the pushed head [a8c492505d3d](https://github.com/peerborne/peerborne/commit/a8c492505d3d07b4be34b9e19b6e86f0f1b6973f). Check invitation keychain freshness; cancel hung provider and ACL waits; reject late aborted worker results; scope fetch limits to bootstrap loads. Relevant regression tests and the applicable build checks passed locally. CI and a fresh Copilot assessment are being rerun on this exact head.

File: packages/core/src/peerborne-document.ts [hunk 7]
```diff
@@ -958,13 +1138,145 @@ export class PeerborneDocument<
...
+    ) => Promise<T>,
+    assertCanApply?: () => void,
+  ): Promise<T> {
+    return this._mutationQueue.run(async () => {
+      if (
+        this._bootstrapLoadApplicationState !== 'pristine' ||
+        this._hashes.size > 0 ||
+        this._lastSyncMessage !== undefined ||
+        this._latestSnapshot !== undefined ||
+        this._subscribed ||
```
[the remark this answers is not in the document]
Comment (robotoer) [reply packages/core/src/peerborne-document.ts:1188]: Addressed in the pushed head [a8c492505d3d](https://github.com/peerborne/peerborne/commit/a8c492505d3d07b4be34b9e19b6e86f0f1b6973f). Check invitation keychain freshness; cancel hung provider and ACL waits; reject late aborted worker results; scope fetch limits to bootstrap loads. Relevant regression tests and the applicable build checks passed locally. CI and a fresh Copilot assessment are being rerun on this exact head.

File: packages/core/src/peerborne-document.ts [hunk 8]
```diff
@@ -1304,73 +1882,152 @@ export class PeerborneDocument<
...
+              : await this._getBlock(
+                  cid,
+                  {
+                    signal: fetchSignal,
+                    maxBlockBytes,
+                    consumeBytes,
+                  },
+                  fetchOptions.getKey,
+                );
+            assertStillActive();
```
[the remark this answers is not in the document]
Comment (robotoer) [reply packages/core/src/peerborne-document.ts]: Addressed in the pushed head [a8c492505d3d](https://github.com/peerborne/peerborne/commit/a8c492505d3d07b4be34b9e19b6e86f0f1b6973f). Check invitation keychain freshness; cancel hung provider and ACL waits; reject late aborted worker results; scope fetch limits to bootstrap loads. Relevant regression tests and the applicable build checks passed locally. CI and a fresh Copilot assessment are being rerun on this exact head.

File: packages/core/src/peerborne-document.ts [hunk 9]
```diff
@@ -1553,20 +2220,28 @@ export class PeerborneDocument<
...
-          );
-        });
+  private async _mergeReaders(
+    changes: ChangesType,
+    assertStillActive?: () => void,
+  ): Promise<void> {
+    await retryACLConflict(() => {
+      assertStillActive?.();
+      return this._readers.merge(changes);
+    });
```
[the remark this answers is not in the document]
Comment (robotoer) [reply packages/core/src/peerborne-document.ts]: Addressed in the pushed head [a8c492505d3d](https://github.com/peerborne/peerborne/commit/a8c492505d3d07b4be34b9e19b6e86f0f1b6973f). Check invitation keychain freshness; cancel hung provider and ACL waits; reject late aborted worker results; scope fetch limits to bootstrap loads. Relevant regression tests and the applicable build checks passed locally. CI and a fresh Copilot assessment are being rerun on this exact head.

File: packages/core/src/peerborne-document.ts [hunk 10]
```diff
@@ -2786,9 +3444,50 @@ export class PeerborneDocument<
...
+    const aggregateBudget: DocumentChangeFetchBudget = {
+      maxBytes: responseLimit,
+      consumedBytes: 0,
+    };
+    const prefetchedBlocks = new Map<string, Uint8Array>();
+    const changeFetchOptions: DocumentChangeFetchOptions<DocumentKey> = {
+      signal,
+      maxBlockBytes: responseLimit,
+      maxAggregateBlockBytes: responseLimit,
+      aggregateBudget,
```
[the remark this answers is not in the document]
Comment (robotoer) [reply packages/core/src/peerborne-document.ts]: Addressed in the pushed head [a8c492505d3d](https://github.com/peerborne/peerborne/commit/a8c492505d3d07b4be34b9e19b6e86f0f1b6973f). Check invitation keychain freshness; cancel hung provider and ACL waits; reject late aborted worker results; scope fetch limits to bootstrap loads. Relevant regression tests and the applicable build checks passed locally. CI and a fresh Copilot assessment are being rerun on this exact head.

File: packages/core/src/peerborne-document.ts [hunk 11]
```diff
@@ -2837,6 +3545,8 @@ export class PeerborneDocument<
         } finally {
           clearResponseDeadline();
         }
+        throwIfLoadAborted(signal);
+        if (!canUseCurrentBootstrapState()) return false;
```
[the remark this answers is not in the document]
Comment (robotoer) [reply packages/core/src/peerborne-document.ts:3556]: Addressed in the pushed head [a8c492505d3d](https://github.com/peerborne/peerborne/commit/a8c492505d3d07b4be34b9e19b6e86f0f1b6973f). Check invitation keychain freshness; cancel hung provider and ACL waits; reject late aborted worker results; scope fetch limits to bootstrap loads. Relevant regression tests and the applicable build checks passed locally. CI and a fresh Copilot assessment are being rerun on this exact head.

File: packages/core/src/peerborne-document-load-boundaries.test.ts [hunk 12]
```diff
@@ -44,31 +60,4215 @@ jest.mock('./peerborne.js', () => ({
...
-    const abort = jest.fn();
+  test('rejects keychain-only state at the queued invitation pristine check', async () => {
+    const apply = jest.fn(async () => undefined);
     const document = fakeDocument({
-      swarm: { config: { loadQuorumTimeoutMs: 1000 } },
+      documentPath: '/invitation-keychain-only',
+      _bootstrapLoadApplicationState: 'pristine',
+      _hashes: new Set(),
+      _subscribed: false,
+      _keychain: { keys: async () => [[new Uint8Array([1]), {}]] },
```
[the remark this answers is not in the document]
Comment (robotoer) [reply packages/core/src/peerborne-document-load-boundaries.test.ts:128]: Keeping this suite together during the stacked bootstrap-hardening changes because its cases share one document harness and deferred-operation model. Focused cases can be selected with Jest -t, and all 99 current load-boundary cases pass. This is a test-organization recommendation rather than a behavior defect; the separate coordinator suite was split in #442. A later extraction should follow the current-only load cleanup so obsolete harness paths are not carried into shared helpers.

Reviewed against published head 78f0832cbb4c30aee578c06ff5a886105f35d678.

File: packages/core/src/peerborne-document.ts [hunk 13]
```diff
@@ -175,13 +177,148 @@ function constantTimeEqual(a: Uint8Array, b: Uint8Array): boolean {
...
 }
 
+const reflectOwnKeys = Reflect.ownKeys;
+
+/** @internal Validate and detach every caller-owned invitation byte field. */
+export function snapshotInvitationBootstrapBundle(
+  bundle: InvitationBootstrapBundle,
+  keyIDLength: number,
+  nonceLength: number,
+): Readonly<InvitationBootstrapBundle> {
```
[the remark this answers is not in the document]
Comment (robotoer) [reply packages/core/src/peerborne-document.ts]: The helper is not re-exported from packages/core/src/index.ts or browser-primitives, and peerborne-document is not an exported package subpath in package.json. Exporting it within the source module permits direct internal tests, but does not add a supported package API. TypeDoc excludes @internal declarations. No public export needs to be removed.

Reviewed against published head 78f0832cbb4c30aee578c06ff5a886105f35d678.

File: packages/core/src/peerborne-document.ts [hunk 14]
```diff
@@ -175,13 +177,148 @@ function constantTimeEqual(a: Uint8Array, b: Uint8Array): boolean {
...
+      throw new TypeError(
+        'Invitation bootstrap bundle must contain exactly its three byte fields',
+      );
+    }
+    recognizedKeys += 1;
+  }
+  if (recognizedKeys !== 3) {
+    throw new TypeError(
+      'Invitation bootstrap bundle must contain exactly its three byte fields',
+    );
```
[the remark this answers is not in the document]
Comment (robotoer) [reply packages/core/src/peerborne-document.ts]: Factored the duplicate invalid-bootstrap-fields message into one internal constant used by both validation branches.

Reviewed against published head 78f0832cbb4c30aee578c06ff5a886105f35d678.

File: packages/core/src/peerborne-document.ts [hunk 15]
```diff
@@ -175,13 +177,149 @@ function constantTimeEqual(a: Uint8Array, b: Uint8Array): boolean {
...
 }
 
+const reflectOwnKeys = Reflect.ownKeys;
+
+/** @internal Validate and detach every caller-owned invitation byte field. */
+export function snapshotInvitationBootstrapBundle(
+  bundle: InvitationBootstrapBundle,
+  keyIDLength: number,
+  nonceLength: number,
+): Readonly<InvitationBootstrapBundle> {
```
[the remark this answers is not in the document]
Comment (robotoer) [reply packages/core/src/peerborne-document.ts]: Moved snapshotInvitationBootstrapBundle into src/internal/invitation-bootstrap.ts and updated the document and test imports. It is no longer exported through peerborne-document or the package barrel. All 97 load-boundary tests and core build pass.

Reviewed against published head 6220478de029be7d22ee325832ee30f872e4b2f3.

File: packages/core/src/peerborne-document.ts [hunk 16]
```diff
@@ -4410,16 +5608,77 @@ export class PeerborneDocument<
...
+            );
+            assertStillActive();
+            beginStateApplication();
+          }
+          assertStillActive();
+          preparedKeychainMerge.commit();
+        } else {
+          beginStateApplication();
+          this._keychain.merge(keychainChanges);
+        }
```
[the remark this answers is not in the document]
Comment (robotoer) [reply packages/core/src/peerborne-document.ts:5674]: A throwing prepared keychain commit now marks bootstrap state pending even when the provider reports a semantic no-op. The new regression fails on the previous code and passes after the fix; all 98 load-boundary tests and the core build pass.

Verified against published head `7d9099ee66c6af8a7c6fe92959cf04860f568907`.

File: packages/core/src/internal/invitation-bootstrap.ts [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+  copyUnsharedUint8Array,
+  snapshotEnumerableOwnDataObject,
+} from '../utils.js';
+
+function assertPositiveSafeByteLimit(value: number, field: string): number {
+  if (!Number.isSafeInteger(value) || value < 1) {
+    throw new RangeError(`${field} must be a positive safe integer`);
+  }
+  return value;
+}
```
[the remark this answers is not in the document]
Comment (robotoer) [reply packages/core/src/internal/invitation-bootstrap.ts]: Moved the positive safe byte-limit validator into one internal module shared by invitation snapshotting and document loading. The focused suite and core build pass.

Verified against published head `7d9099ee66c6af8a7c6fe92959cf04860f568907`.

File: packages/core/src/peerborne-document.ts [hunk 18]
```diff
@@ -1036,20 +1407,162 @@ export class PeerborneDocument<
...
+          `Failed to prepare remote update notification for ${this.documentPath}`,
+        );
+      })
+      .finally(() => {
+        if (this._remoteUpdateNotificationTail === queued) {
+          this._remoteUpdateNotificationTail = undefined;
+        }
+      });
+    this._remoteUpdateNotificationTail = queued;
+  }
```
[the remark this answers is not in the document]
Comment (robotoer) [reply packages/core/src/peerborne-document.ts:1485]: The stored tail already catches task failures before it becomes the next predecessor. Added a regression that queues two notifications, fails the first audience lookup, and proves the second dispatches once. The load-boundary and catch-up suites pass.

Verified against published head `966952208b7513311dd877874973ea3be5003586`.

File: packages/core/src/peerborne-document.ts [hunk 19]
```diff
@@ -1036,20 +1407,162 @@ export class PeerborneDocument<
...
+      this._writers.users(),
+    ]);
+    // ACL implementations may reject overlapping operations, so keep at most
+    // one read against the reader ACL in flight.
+    const filteredWriters: PublicKey[] = [];
+    for (const writer of writers) {
+      if ((await this._readers.check(writer)) !== true) {
+        filteredWriters.push(writer);
+      }
+    }
```
[the remark this answers is not in the document]
Comment (robotoer) [reply packages/core/src/peerborne-document.ts:1457]: Reader ACL operations deliberately run one at a time: supported opaque ACL providers reject overlapping reads, and their membership check owns identity equivalence. Replacing that check with ad-hoc object/string identity or concurrent calls would break the provider contract. The FIFO contention/retry tests cover this ordering.

Verified against published head `966952208b7513311dd877874973ea3be5003586`.

File: packages/automerge/src/bootstrap-keychain-load.test.ts [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+        false,
+        beginStateApplication,
+        false,
+        logicalKeychainChange,
+      ),
+    ).resolves.toBe(true);
+    expect(beginStateApplication).toHaveBeenCalledTimes(1);
+    expect(logicalKeychainChange).toHaveBeenCalledTimes(1);
+    expect(target.getKey(keyId)).toBeDefined();
+    await expect(target.keys()).resolves.toHaveLength(1);
```
[the remark this answers is not in the document]
Comment (robotoer) [reply packages/automerge/src/bootstrap-keychain-load.test.ts:76]: Kept separate adapter integration fixtures because each must prove its own keychain hydration and wire representation through the document boundary. Sharing the body across packages would couple otherwise independent provider evidence; both suites remain exercised by the repository test command.

Verified against published head `966952208b7513311dd877874973ea3be5003586`.

File: packages/core/src/invitation-catch-up.ts [hunk 21]
```diff
@@ -176,12 +176,12 @@ export async function withIssuerPinnedInvitationStream<
...
   founderAddress: string,
   dial: (address: string, signal: AbortSignal) => Promise<Stream>,
-  loadAndVerify: (stream: Stream) => Promise<Result>,
+  loadAndVerify: (stream: Stream, signal: AbortSignal) => Promise<Result>,
   timeoutMs: number = INVITATION_STREAM_TIMEOUT_MS,
 ): Promise<Result> {
   return withInvitationProtocolStream(
     (signal) => dial(founderAddress, signal),
-    (stream) => loadAndVerify(stream),
+    (stream, signal) => loadAndVerify(stream, signal),
```
[the remark this answers is not in the document]
Comment (robotoer) [reply packages/core/src/invitation-catch-up.ts]: Passed loadAndVerify directly to withInvitationProtocolStream, removing the redundant forwarding closure.

Verified against published head `966952208b7513311dd877874973ea3be5003586`.
